objects/src/transaction/proven_tx.rs

```diff
@@ -13,14 +13,107 @@ use crate::{
 // ================================================================================================
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub enum AccountDetails {
-    /// The whole state is needed for new accounts
-    Full(Account),
+pub enum AccountUpdateDetails {
+    /// Account is private (no on-chain state change).
+    Private,
+
+    /// The whole state is needed for new accounts.
+    New(Account),
 
     /// For existing accounts, only the delta is needed.
     Delta(AccountDelta),
 }
 
+impl AccountUpdateDetails {
+    /// Returns `true` if the account update details are for private account.
+    pub fn is_private(&self) -> bool {
+        matches!(self, Self::Private)
+    }
+}
+
+/// Account update data.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AccountUpdate {
+    /// The hash of the account before the transaction was executed.
+    ///
+    /// Set to `Digest::default()` for new accounts.
+    init_hash: Digest,
+
+    /// The hash of the account after the transaction was executed.
+    final_hash: Digest,
+
+    /// Optional account state changes used for on-chain accounts. This data is used to update an
+    /// on-chain account's state after a local transaction execution. For private accounts, this
+    /// is [AccountUpdateDetails::Private].
+    details: AccountUpdateDetails,
+}
+
+impl AccountUpdate {
+    /// Creates a new [AccountUpdate].
+    pub const fn new(init_hash: Digest, final_hash: Digest, details: AccountUpdateDetails) -> Self {
+        Self { init_hash, final_hash, details }
+    }
+
+    /// Returns the initial account state hash.
+    pub fn init_hash(&self) -> Digest {
+        self.init_hash
+    }
+
+    /// Returns the final account state hash.
+    pub fn final_hash(&self) -> Digest {
+        self.final_hash
+    }
+
+    /// Returns the account update details.
+    pub fn details(&self) -> &AccountUpdateDetails {
+        &self.details
+    }
+
+    /// Returns `true` if the account update details are for private account.
+    pub fn is_private(&self) -> bool {
+        self.details.is_private()
+    }
+}
+
+/// Account update data.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AccountUpdateData {
+    /// Account ID.
+    account_id: AccountId,
+
+    /// The hash of the account after the transaction was executed.
+    final_state_hash: Digest,
+
+    /// Account update details.
+    details: AccountUpdateDetails,
+}
```

Comment:
It seems like we have 2 structs which are very similar: `AccountUpdate` and `AccountUpdateData`. These look as follows:
```Rust
pub struct AccountUpdate {
    init_hash: Digest,
    final_hash: Digest,
    details: AccountUpdateDetails,
}

pub struct AccountUpdateData {
    account_id: AccountId,
    final_state_hash: Digest,
    details: AccountUpdateDetails,
}
```
I think having these two structs being so similar is a bit confusing - but I'm not sure what's the best solution here.

Ideally, we'd have just a single struct which looks something like this:
```Rust
pub struct AccountUpdate {
    account_id: AccountId,
    new_state_hash: Digest,
    details: AccountUpdateDetails,
}
```
This would replace `AccountUpdateData` - but it is missing `init_hash` to make it useful in the `ProvenTransaction`. Potential options are:

1. Go back to having individual fields (i.e., `init_account_hash`, `final_account_hash` etc.) in `ProvenTransaction`.
2. Use both `initi_account_hash` and `account_update` in `ProvenTransaction`.
3. Wrap `AccountUpdate` in another struct which has `init_account_hash`.

For option 2, the `ProvenTransaction` struct would look something like this:
```Rust
pub struct ProvenTransaction {
    id: TransactionId,
    init_account_hash: Digest,
    account_update: AccountUpdate,
    input_notes: InputNotes<Nullifier>,
    output_notes: OutputNotes,
    block_ref: Digest,
    proof: ExecutionProof,
}
```
This looks a bit weird to me - so, I'm not sure it's a good option.

For option 3, we could have something like this:
```Rust
pub struct AccountUpdateInfo {
    init_state_hash: Digest,
    update: AccountUpdate,
}

impl AccountUpdateInfo {
    pub fn init_hash(&self) -> AccountId {
        self.update.account_id
    }

    pub fn init_hash(&self) -> Digest {
        self.init_hash
    }

    pub fn final_hash(&self) -> Digest {
        self.update.final_hash
    }

    pub fn details(&self) -> &AccountUpdateDetails {
        &self.update.details
    }
}

pub struct ProvenTransaction {
    id: TransactionId,
    account_update: AccountUpdate,
    input_notes: InputNotes<Nullifier>,
    output_notes: OutputNotes,
    block_ref: Digest,
    proof: ExecutionProof,
}

impl ProvenTransaction {
    pub fn account_id(&self) -> AccountId {
        self.account_update.account_id()
    }
}
```

I think this looks much better - but I don't love the `AccountUpdateInfo` name.

So, I guess the question is whether we want to go with option 1, option 3, or maybe something else?